### ISSUE TITLE
Onboarding Existing Node-RED's - FlowFuse Plugin

### ISSUE THREAD

**Issue by joepavitt** [FlowFuse/flowfuse#4914]

**Comment by gstout52** [comment]
@joepavitt Can you provide an estimate on how much design work is needed here? We can deprioritize other 2.18 items to deliver on this if needed.

**Comment by joepavitt** [comment]
This is mostly with @knolleary on sizing. It was more engineering design than UX design work because we weren't sure on the scope of what's possible to control via a plugin

The onboarding existing NR has got better because of the device agent improvements shipped in 2.17 too which can auto import existing NR flows

**Comment by gstout52** [comment]
@joepavitt I'm conceiving of a revised onboarding flow that begins by asking the new user if they have Node-RED instances running elsewhere that they'd like to connect to, and if so, where they're hosted and then connect them to FlowFuse. With the Device Agent improvements just shipped, would this piece of work provided the remaining engineering work to permit us to offer that kind of onboarding experience?

**Comment by joepavitt** [comment]
So, for Hosted Instances, it's straight forward. For Remote, it's more complicated.

For the new device agent changes, we have the ability to import any flows found on the same hardware, but they still need to do a manual install of the device agent.

An alternative route would be a FlowFuse plugin for Node-RED which the user would install in their own local Node-RED instance, and that in turn would connect to FlowFuse and permit saving of snapshots. The challenge here is though that it doesn't permit replication of the full "Device" workflow, we cannot get remote access, restart Node-RED, etc from within a NR plugin.

Asking them where they are hosted is fine, but currently, we have no way of automatically connecting to that. The user would need to install device agent in that place

**Comment by gstout52** [comment]
@joepavitt Thank you. Does this sound right?

A user has an instance of Node-RED running somewhere.

Either: (A) it is hosted in a cloud service or on other hardware, or (B) it is running on an edge device.

If (A), then "it's straightforward" (quoting you from above). I guess we'd just ask for where it is hosted and connect it?
If (B), then either:
- (B1) the user needs to install the Device Agent on the device, which will import the Node-RED instance running there and connect to FlowFuse [better] or 
- (B2) this new plugin can be installed into the remotely running Node-RED which will connect it to FlowFuse but allow for more limited control within FlowFuse [worse but better than what we have now].

**Comment by knolleary** [comment]
If the user wants to migrate an existing NR application into the FlowFuse platform as a *hosted instance*, they can use the NR Tools Plugin to create a snapshot on their existing NR and push it to a hosted instance in FF. They then shutdown that external NR instance. I think that's what (A) is referring to - and is covered in the docs.

If the user wants to connect an existing NR as a Remote Instance, then they need to use the Device Agent. (B1)

The concept of this issue was to implement a plugin to do B2. I had looked at this in detail a while ago - I'm sure I wrote up my thoughts/findings on it, but I can't find them now.

We also discussed this last week. The *hard* part of implementing this plugin is it requires a new type of 'thing' on the platform. Our existing model of a device assumes a lot of basic functionality that the plugin won't have. So we need to go through every single piece of device functionality in both backend and frontend, and add 'awareness' of this new type of device so we don't offer-up bits of functionality we can't provide. When I started a quick PoC of the plugin, I quickly hit this and it was an overwhelming amount of change that made the PoC no-longer quick nor sensible to continue at the time.

I will put some more effort into a sizing here for the plugin - but not quick.

**Comment by gstout52** [comment]
After further discussion with Nick and Joe, I'm taking the milestone off of this one and putting it to the backlog.

**Comment by dgatti0213** [comment]
Jumping in from the SE side. We talked through most of this in a huddle today, so this is mostly to capture the conclusion on the issue.

There's a premise underneath the A / B2 paths worth stating: both assume the flows can run on a hosted instance, which only holds if the flows aren't hardware-dependent. For a large and growing slice of the installs we're trying to onboard, they are. Node-RED is on the device precisely because the flows touch something local: serial, GPIO, fieldbus, an IO-Link master's ports (IFM), the GX D-Bus (Victron), local sensors over USB. We hit this exact wall on a Phidgets SBC4, cloud-hosted FlowFuse couldn't reach the local sensors, so the agent had to run on the device. Snapshot those flows to a hosted instance and they don't run.

So for these, "migrate off the device" isn't onboarding, it's breakage. The need is the opposite: manage the flows that have to stay on the device. That's normally B1 (Device Agent), but on the hardware these flows live on (Tulip Edge, Victron Venus OS, IFM AL-series, Opto groov), installing the agent is itself the blocker.

Where we landed in the huddle: the in-place plugin (B2) is the right step here, scoped deliberately small. Not migration, and not full Device parity. On this hardware the OEM already owns the NR process, so the plugin's main limitation (no remote stop/start, no settings ownership) matches reality rather than compromising it. The job is "get the flows under FF management": snapshot, deploy, and version flows, with deploys pushing a flow set and triggering a runtime reload via the Admin API. Remote lifecycle and the editor proxy are explicit non-goals for v1.

@knolleary is putting together the list of what's feasible so we can evaluate whether there's an MVP worth taking to market.

**Comment by knolleary** [comment]
@dgatti0213 the sizing is for what engineering effort would be needed to solve this in the way you've described.

>  the in-place plugin (B2) assume the flows can run on a hosted instance. 

That is not the case. B2 is about having a plugin for an *existing* node-red instance that allows FF to manage it - I believe that's what you're looking for. It is not about migrating flows anywhere else.

The core problem is the life-cycle expectation of a Remote Instance cannot be satisfied by a plugin running inside Node-RED. It requires a process external to Node-RED that knows how to stop/start NR. To provide the security model and integration with FF (such as the remote editor proxy) it needs to be able to modify the settings file. If we don't own the NR process/settings file, these aren't things we can easily do.

This is why B2 describes itself as having more limited functionality as to what the Device Agent allows - no remote stop/start, no remote editing flows. At best, the ability to snapshot flows remotely. This is why we pulled back from this approach at the time - the engineering cost to make the platform aware of this new limited type of device and the plugin itself.

**Comment by knolleary** [comment]
I have updated the description to give more structure to this headline item. Next task will be to create subissues for the work involved that are finer grained. These will be split between this repo (for platform work) and the repo created for the plugin.

**Comment by dgatti0213** [comment]
When you are ready, I have some hardware from various vendors we can test on and I can get a few more from others that are interested in partnership.

**Comment by ZJvandeWeg** [comment]
Early this discussion mentioned the NR Tools Plugin: https://github.com/FlowFuse/flowfuse/issues/4914#issuecomment-2872738808, then there's also the `nr-assistant` plugin. Much of the same plumbing, sign in, API calls to FlowFuse, etc. Now we're proposing a new plugin, that does many of the same API calls.

Further, there's a lot platform work identified: https://github.com/FlowFuse/flowfuse/issues/4914#issuecomment-4682649757, though in my mind incorrectly sequenced. From day 1 this plugin should have customer value, and ideally (but also to make this worth time investment) the sequence is reverted right away. In one release we **must** release 2 or maybe 3 versions and adapt to installation count, much like done when Dashboard 2.0 was build. Similar, we should aim for _hundreds_ of installs/users in 2 months.

How, though? We merge `nr-tools-plugin` and `nr-assistant`. We don't lead with making a 'device'-light. The B2 route here is dependant on showing value from the second that a plugin is installed. Not days or weeks later. It's my fear when we're going 'hygiene' route again with better 'management'. There's only value in management if FlowFuse helps building the first production application.

@knolleary My opinion but also strong conviction the roadmap is:

| Week | Tasks | Signal |
|--------|--------|--------|
| 1 | <ol><li>Rename `nr-assistant` to what this plugin will be called. Reduce plugin footprint of FF</li><li>publish v0.1 of a plugin</li><li>publish in NR community about new direction</li><li>sign ups are attributed to this plugin in Hubspot (create campaign etc)</li><li>Set up meetings with HW partners</li></ol> | Downloads, verified attributable FF accounts |
| 2 | <ol><li>Extend AI assistant capabilities</li><li>Set up Auth for Node-RED, not just sign in for FF</li><li>publish v0.2 of a plugin</li><li>Community outreach, AI + NR = FF</li><li>Internal SDR campaign starts with registrations</li></ol> | Downloads / Usage is growing 50% WoW, API-calls to NR |
| 3 | <ol><li>Merge capabilities of nr-tools</li><li>Stream logs to Remote instance overview</li><li>Best of class docs for the plugin, CTA on website.</li><li>Community stream to explain the plugin (outreach)</li></ol> | Cell |
| 4 | <ol><li>Verify on all hardware partners that have shown interest</li><li>MCP platform capabilities in plugin</li><ol> | Still at 50% WoW growth on downloads, meaningful new sign ups every week |

Then onwards from there.

**Comment by dimitrieh** [comment]
> then there's also the nr-assistant plugin. Much of the same plumbing, sign in, API calls to FlowFuse, etc.

I agree in this, that the reuse of the same plugin would be welcome so there is only one plugin to get connected to FlowFuse, rather than multiple.

---

> when a user deploys new flows, generate a snapshot and save to the platform

I am curious if for these devices if the assumption is that these are just devices to be managed/deployed to or that they actually also should be the control plane, meaning that users would connect to them directly and edit things there in the editor. If they are merely deployment targets this might change the situation here? (meaning we could continue offer the editing experience in a hosted instance for example).

**Comment by knolleary** [comment]
I've discussed some of this with individuals separately, but need to get it written down; keeping in mind this is a public issue so not all context can be shared.

There are two different goals being expressed here; both valid, but with different consequences in terms of actions.

1. (the original concept) A NR plugin to get *existing* NR installs connected as a 'remote' instances on FF. This is primarily targeted to the device manufacturers that preinstall Node-RED. Drew has been in conversation with a number of them and believes this will unlock commercial opportunities.
2. A more community-oriented play - a plugin that gives instant value, starting with nr-assistant and iterate from there.

They are both valid goals; but there is a tension between them in terms of priority and resource.

1. The nr-assistant plugin *already* allows a user to sign-in to FFC in order to enable it in a standalone NR installed. It connects and authenticates as a *user* on FFC.
   - the nr-assistant feature set is much smaller than the full FF Expert feature set. 
   - We didn't do much marketing around the release of this - at least, nothing beyond the announcement and initial material. It has not seen any adoption as it has dropped out of the conversation before we really began.

2. The onboarding plugin would connect as a remote instance - albeit with a small subset of functionality initially. Within the FF UI, it would should up as a remote instance.
   - Remote instances are only available to Edge/Fleet customers (which aligns with the target audience through the device partners)
   - This would *include* ff-assistant enablement


My proposal is:

1. The ff-assistant plugin already provides a way for unmanaged NRs to install *something* to start getting value from FF. We don't need to reinvent that. The work here is more marketing oriented:
    - Work on a marketing plan to promote it and reinvigorate interest around it
    - Ensure we have a way to track adoption

2. Work on the Device Agent plugin - it unlocks value for device partners that we have concrete feedback on. We plan on doing a partner session at NR Con with one of the device manufacturers - having this plugin available will be key part of the story we can tell at the event. This is the engineering focussed work in the short term.

**Comment by ZJvandeWeg** [comment]
Any update on progress for this week Nick? We're halfway in the release and it's not clear what will be delivered when.

**Comment by knolleary** [comment]
@ZJvandeWeg as per my daily standup status, progress has been blocked this week due to combination of Node-RED security issues that needed timely resolution, customer support issues and unblocking Esme with issues around the new packages on FFC - we have also been short-handed this week so not been able to delegate as I normally would. That is all pretty much complete now, so full focus on the plugin work from here on.
